File: apps/backend/app/core/auth/auth_failure.py
"""401 auth-failure response helper.

When the backend rejects a request because the session is dead (idle,
expired, revoked, or never existed) we want two things on the way out:

1. **Standardized JSON shape** — `{"error": "<reason>"}` where reason is
   one of `session_idle_expired` / `session_expired` / `unauthenticated`,
   so the SPA can map to a banner explaining why the user is back at
   `/login`.
2. **Clear the stale cookies** — the next request from the browser
   starts without `yaaos_session` / `yaaos_csrf` so it can't keep hitting
   the same 401 loop and cascading "Not signed in" errors across pages.

`AuthFailure` subclasses `HTTPException` so FastAPI's default handler
still produces a sane 401 if the custom handler isn't registered (test
apps that bypass `create_app`). `register_handler(app)` — called from
the app factory — adds the Set-Cookie clears.

`auth_failure_response()` is the equivalent for endpoints that return a
`Response` directly rather than raising (e.g. `/api/auth/me`, which is a
public route so it can't reach the dependency-side raise path).
"""

from __future__ import annotations

from typing import Literal

from fastapi import FastAPI, HTTPException, Request
from fastapi.responses import JSONResponse

from app.core.auth.cookies import CSRF_COOKIE_NAME, SESSION_COOKIE_NAME, clear_cookie_attrs

AuthFailureReason = Literal[
    "session_idle_expired",
    "session_expired",
    "unauthenticated",
]


class AuthFailure(HTTPException):
    """401 caused by a dead/missing session. Raise from FastAPI deps —
    the registered handler clears `yaaos_session` + `yaaos_csrf` on the
    way out so the browser's next request starts clean."""

    def __init__(self, reason: AuthFailureReason) -> None:
        super().__init__(status_code=401, detail={"error": reason})
        self.reason: str = reason


def auth_failure_response(reason: AuthFailureReason) -> JSONResponse:
    """JSONResponse equivalent for endpoints returning Response directly
    (no raise). Same body shape + Set-Cookie clears as the exception-
    handler path."""
    resp = JSONResponse(status_code=401, content={"error": reason})
    resp.set_cookie(**clear_cookie_attrs(SESSION_COOKIE_NAME))
    resp.set_cookie(**clear_cookie_attrs(CSRF_COOKIE_NAME))
    return resp


def register_handler(app: FastAPI) -> None:
    """Wire `AuthFailure` → 401 with cleared cookies. Called from
    `create_app`; tests that need the cookie behavior call it on their
    test app."""

    @app.exception_handler(AuthFailure)
    async def _handler(_: Request, exc: AuthFailure) -> JSONResponse:
        return auth_failure_response(exc.reason)  # type: ignore[arg-type]


__all__ = [
    "AuthFailure",
    "AuthFailureReason",
    "auth_failure_response",
    "register_handler",
]

File: apps/web/src/core/api/auth-failure.ts
/**
 * Central 401 handler for `apiFetch`.
 *
 * When the backend rejects a request as auth-dead (idle timeout, hard
 * expiry, or no session at all), it returns `{"error": "<reason>"}` with
 * `Set-Cookie` clears for `yaaos_session` + `yaaos_csrf` already on the
 * response. This module takes the next two steps the browser needs:
 *
 *  1. **One-time hard redirect to `/login`** with a `?reason=` query for
 *     the banner copy and a `?next=` query carrying the current path so
 *     the user lands back where they were trying to go after sign-in.
 *     `window.location.assign` (not soft-nav) is intentional — it clears
 *     React state + TanStack Query cache so stale "logged in" data
 *     doesn't keep triggering 401 cascades.
 *
 *  2. **Mutex so concurrent 401s redirect once.** Every page often fires
 *     2-3 queries in parallel (`/api/auth/me`, `/api/orgs/mine`,
 *     `/api/notifications/popover`, page-specific queries). Without a
 *     mutex, each one would call `assign` and the URL bar would race.
 *
 * `next` round-trips through `LoginPage`, which forwards it as a
 * `next=` query on `/api/auth/login?provider=...&next=...`. The backend's
 * `_safe_next` validator in `apps/backend/app/domain/sessions/web.py`
 * already rejects open-redirect attempts (must start with single `/`,
 * collapses anything else to `/`). We mirror that allowlist client-side
 * so `next` is sane before it ever hits the URL.
 */

export type AuthFailureReason = "expired" | "idle" | "signed_out";

/** Thrown from `apiFetch` after the redirect is triggered. React Query
 * sets `isError`/`data: undefined` on the calling hook; no page-level
 * component should branch on this, because the browser is about to
 * navigate away anyway. */
export class AuthError extends Error {
  constructor(public readonly reason: AuthFailureReason) {
    super(`auth_failure:${reason}`);
    this.name = "AuthError";
  }
}

/** Same-origin path allowlist mirroring backend `_safe_next`. Returns
 * the input if it's a safe relative path; null otherwise. Exported so
 * the LoginPage can validate `next` on submit without re-stating the
 * rule. */
export function safeNext(value: string | null | undefined): string | null {
  if (!value) return null;
  // Single leading `/`, no `//evil.com`, no `\\`, no scheme-relative.
  if (!value.startsWith("/")) return null;
  if (value.startsWith("//")) return null;
  if (value.includes("\\")) return null;
  // Don't loop the user back to an auth path.
  if (value === "/login" || value.startsWith("/login?") || value.startsWith("/login/")) return null;
  if (value === "/logout" || value.startsWith("/logout?") || value.startsWith("/logout/"))
    return null;
  return value;
}

/** Map the backend `{"error": <code>}` body to the UX reason the banner
 * keys off of. Backend → UX:
 *   `session_idle_expired` → `idle`
 *   `session_expired`      → `expired`
 *   `unauthenticated`      → `signed_out`
 *   anything else          → `signed_out` (catch-all so a renamed code
 *                            still produces a banner). */
export function reasonFromBody(body: unknown): AuthFailureReason {
  if (typeof body === "object" && body !== null && "error" in body) {
    const err = (body as { error: unknown }).error;
    if (err === "session_idle_expired") return "idle";
    if (err === "session_expired") return "expired";
    if (err === "unauthenticated") return "signed_out";
  }
  return "signed_out";
}

let redirectInProgress = false;

/** Test hook. Resets the once-per-page-load mutex so each `it` starts
 * with a clean state. NOT for production use. */
export function _resetAuthFailureForTests(): void {
  redirectInProgress = false;
}

/** Called from `apiFetch` on a 401. Reads the body, captures the
 * current path as `next`, hard-navigates to `/login?reason=...&next=...`.
 * Idempotent across concurrent 401s. Throws `AuthError` so callers stop
 * using the (empty) response body. */
export async function handleAuthFailure(resp: Response): Promise<never> {
  let body: unknown = null;
  try {
    body = await resp.json();
  } catch {
    // Body wasn't JSON (older endpoint, network glitch). Fall through —
    // reasonFromBody returns "signed_out" as the catch-all.
  }
  const reason = reasonFromBody(body);

  if (!redirectInProgress) {
    redirectInProgress = true;
    const here = window.location.pathname + window.location.search + window.location.hash;
    const next = safeNext(here);
    const url = new URL("/login", window.location.origin);
    url.searchParams.set("reason", reason);
    if (next) url.searchParams.set("next", next);
    window.location.assign(url.toString());
  }

  throw new AuthError(reason);
}
